guides/micronaut-kotlin-extension-fns/micronaut-kotlin-extension-fns.adoc

```diff
@@ -0,0 +1,173 @@
+    val tmp = this[index1] // 'this' corresponds to the list
+    this[index1] = this[index2]
+    this[index2] = tmp
+}
+----
+
+Then use it like it was a regular method of `MutableList`
+
+[source, kotlin]
+----
+val languages = mutableListOf('java', 'groovy', 'kotlin')
+
+languages.swap(0,2) // Will swap 'java' and 'kotlin' so kotlin comes first
+----
+
+=== Using Extension Functions in an App
+
+To show off the extension functions available for use in Micronaut (and how to write your own), we'll build a simple application that
+
+- Consumes https://icanhazdadjoke.com/api[this fun dad joke API] (I love a good dad joke) with an http client
+- Schedules the joke to "be sent" to someone [we won't actually be integrating a message sender for simplicity]
+- Writes our own extension function for the client (as if it was provided by a third party)
+- Puts everything together in a controller
+- Starts the application with the `startApplication` extension function
+
+include::{commondir}/common-create-app.adoc[]
+
+=== The Micronaut Kotlin Extension Functions
+
+To aid kotlin developers in writing more idiomatic kotlin code in Micronaut, we have provided a
+https://micronaut-projects.github.io/micronaut-kotlin/latest/guide/index.html#extensionFunctions[kotlin extension function library]
+for use.
+
+To use them in our project, we simply need to include the dependency
+
+dependency:micronaut-kotlin-extension-functions[groupId=io.micronaut.kotlin]
+
+You can see the https://micronaut-projects.github.io/micronaut-kotlin/api/micronaut-kotlin-extension-functions[dokka docs for the full list of functions here]
+
+Big credit is due to https://github.com/ideaplugins[Alejandro Gomez] for allowing for the use of his initial collection of extension functions!
+
+Let's see these functions in action!
+
+=== Application
+
+Right away let's modify the generated `Application.kt` to use the `startApplication` extension function
+
+source:Application[]
+
+<1> Here we import the extension function
+<2> Since the extension function takes a type argument, we'll define our `Application` as a kotlin `object`, which is an easy shortcut for a singleton or place to put a static method like `main`
+<3> `startApplication<Application>` here does the work of `.build().mainClass(Application::class.java).start()`. Convenient!
+
+=== DadJokeClient
+
+Next we'll build out our functionality. We'll start by modelling the response types for the plain `GET` request for the Dad Joke API (that just return a random joke),
+as well as the paged results that are returned from the `/search` endpoint.
+
+source:DadJoke[]
+
+source:DadJokePagedResults[]
+
+Then, we can create a standard Micronaut http client for the DadJoke API endpoint, letting the `@Client` annotation implement the interface we lay out.
+
+source:DadJokeClient[]
+
+Note that while *we* are creating this client for the purposes of this guide, but often times something like this may come from
```

Comment:
The combination of `while` and `but` reads a little awkward.  Is it better if we delete 1 of those 2 words, or leave it as it is?